File: elixir/lib/word_count.ex
defmodule WordCount do
  @moduledoc false

  @default_max_word 64
  @max_word 1024
  @min_word 4

  def count_bytes(bytes, top, max_word) when is_binary(bytes) do
    max_word = normalize_max_word(max_word)

    {counts, total, word, _stored, in_word} =
      bytes
      |> :binary.bin_to_list()
      |> Enum.reduce({%{}, 0, [], 0, false}, &step(&1, &2, max_word))

    {counts, total} =
      if in_word do
        finish_word(counts, total, word)
      else
        {counts, total}
      end

    entries =
      counts
      |> Enum.map(fn {word, count} -> %{word: word, count: count} end)
      |> Enum.sort_by(fn entry -> {-entry.count, entry.word} end)
      |> Enum.take(top)

    %{total: total, unique: map_size(counts), top: entries}
  end

  defp step(byte, {counts, total, word, stored, in_word}, max_word) do
    cond do
      letter?(byte) ->
        if stored < max_word do
          {counts, total, [lower_ascii(byte) | word], stored + 1, true}
        else
          {counts, total, word, stored, true}
        end

      in_word ->
        {counts, total} = finish_word(counts, total, word)
        {counts, total, [], 0, false}

      true ->
        {counts, total, word, stored, false}
    end
  end

  defp finish_word(counts, total, word) do
    key = word |> Enum.reverse() |> List.to_string()
    {Map.update(counts, key, 1, &(&1 + 1)), total + 1}
  end

  defp letter?(byte) do
    lower = Bitwise.bor(byte, 32)
    lower >= ?a and lower <= ?z
  end

  defp lower_ascii(byte) when byte >= ?A and byte <= ?Z, do: byte + 32
  defp lower_ascii(byte), do: byte

  defp normalize_max_word(0), do: @default_max_word
  defp normalize_max_word(max_word) when max_word < @min_word, do: @min_word
  defp normalize_max_word(max_word) when max_word > @max_word, do: @max_word
  defp normalize_max_word(max_word), do: max_word
end

File: lua/src/wordcount.lua
local M = {}

local ORACLE_DEFAULT_MAX_WORD = 64
local MAX_WORD = 1024
local MIN_WORD = 4

local function normalize_max_word(value)
  if value == 0 then
    return ORACLE_DEFAULT_MAX_WORD
  end
  return math.min(math.max(value, MIN_WORD), MAX_WORD)
end

local function ranks_before(left, right)
  if left.count ~= right.count then
    return left.count > right.count
  end
  return left.word < right.word
end

local function rank_counts(counts, top)
  local entries = {}
  for word, count in pairs(counts) do
    entries[#entries + 1] = { word = word, count = count }
  end
  table.sort(entries, ranks_before)

  local limited = {}
  for index = 1, math.min(top, #entries) do
    limited[index] = entries[index]
  end
  return limited, #entries
end

function M.count_bytes(bytes, top, max_word)
  max_word = normalize_max_word(max_word)

  local counts = {}
  local word = {}
  local stored = 0
  local total = 0

  for index = 1, #bytes do
    local byte = string.byte(bytes, index)
    local lower = byte | 32
    if lower >= 97 and lower <= 122 then
      if stored < max_word then
        stored = stored + 1
        word[stored] = string.char(lower)
      end
    elseif stored > 0 then
      local key = table.concat(word, "", 1, stored)
      counts[key] = (counts[key] or 0) + 1
      total = total + 1
      word = {}
      stored = 0
    end
  end

  if stored > 0 then
    local key = table.concat(word, "", 1, stored)
    counts[key] = (counts[key] or 0) + 1
    total = total + 1
  end

  local entries, unique = rank_counts(counts, top)
  return { total = total, unique = unique, top = entries }
end

return M
